### ISSUE TITLE
jax.grad does not recognize jax.numpy.nan_to_num 

### ISSUE THREAD

**Issue by sjiang23** [google/jax#8247]
Hi JAX team,

I come across the following error:

```
def foo(x):
  return jnp.sum(jnp.nan_to_num(x/jnp.linspace(-1,1,21), posinf = 0))

foo(2.)
DeviceArray(-2.5749207e-05, dtype=float32)

grad(foo)(2.)
DeviceArray(nan, dtype=float32, weak_type=True)
```

`grad(foo)` should recognize the `inf` in `x/linspace(-1,1,21)` has been set as 0 by `nan_to_num`?

Thanks for answering.

**Comment by patrick-kidger** [comment]
This is a limitation of autodifferentiation techniques in general. c.f. #5039 and #1052.

In this case the trick is to avoid ever creating a `nan` in the first place:
```python
def foo(x):
    div = jnp.linspace(-1, 1, 21)
    div = jnp.where(div == 0, jnp.inf, div)
    return jnp.sum(x/div)
```

**Comment by jakevdp** [comment]
You can also understand this symbolically; say you have a function *y = f(x / a)*. The derivative of this is *y' = (1/a) f'(x / a)*. Now let *f = nan_to_num* and *a = 0*; it's easy to see that regardless of the output *f'* , the gradient will still be infinite.

To fix this, I'd suggest exactly what @patrick-kidger suggested.

**Comment by sjiang23** [comment]
@patrick-kidger  @jakevdp Thanks alot. I will make changes as suggested.
